torch/_functorch/_aot_autograd/schemas.py

```diff
@@ -171,47 +171,56 @@ class SubclassCreationMeta:
     flat_tensor_start_idx: int
     # The number of tensors that live in this subclass wrapper
     arg_count: int
+    # meta and inner_keys are produced by the subclass's __tensor_flatten__.
+    # We need to keep them around along with outer_size / outer_stride to plumb them
+    # into __tensor_unflatten__
+    attrs: Dict[str, Union["SubclassCreationMeta", None]]
+    outer_size: List[int]
+    outer_stride: List[int]
+    meta: Any
     # Stores the original subclass itself.
     # This is needed because we need the autograd metadata on the original subclass
     # (this is guaranteed to be a wrapper subclass that holds a fake tensor,
     #  so holding onto this at runtime shouldn't leak memory)
-    original_subclass: torch.Tensor
-    # meta and inner_keys are produced by the subclass's __tensor_flatten__.
-    # We need to keep them around along with outer_size / outer_stride to plumb them
-    # into __tensor_unflatten__.
-    meta: Any
-    inner_keys: List[Any]
-    outer_size: Tuple[int, ...]
-    outer_stride: Tuple[int, ...]
+    original_subclass: Any
 
     def creation_fn(self, all_args, *, is_runtime: bool):
-        curr_args = all_args[
-            self.flat_tensor_start_idx : self.flat_tensor_start_idx + self.arg_count
-        ]
-        assert len(curr_args) == len(
-            self.inner_keys
-        ), f"inner_keys: {str(self.inner_keys)}. len(curr_args): {len(curr_args)}"
-        # NB: Sometimes we have real inner tensors and symbolic metadata.
-        # TODO: Resolve this so we always have matching real / symbolic tensors / metadata.
-        out = type(self.original_subclass).__tensor_unflatten__(  # type: ignore[attr-defined]
-            dict(zip(self.inner_keys, curr_args)),
-            self.meta,
-            self.outer_size,
-            self.outer_stride,
+        inner_tensors = {}
+
+        curr_start_idx = self.flat_tensor_start_idx
+        for attr, creation_meta in self.attrs.items():
+            if creation_meta is None:
+                subclass = all_args[curr_start_idx]
+                curr_start_idx += 1
+            else:
+                subclass = creation_meta.creation_fn(all_args, is_runtime=is_runtime)
+                curr_start_idx += creation_meta.arg_count
+            inner_tensors[attr] = subclass
+
+        rebuilt = type(self.original_subclass).__tensor_unflatten__(
```

Comment:
All of the indices in this reconstruction are definitely non-trivial. It would be great if we had some runtime debug-asserts we could run that would tell us if we messed up the indexing somewhere, so we get a less cryptic error if we get this wrong 🤔. I can't think of a great way to do this though, unless we do something like save all of the shapes of the inner tensors at trace time and assert that our reconstructed inner tensors are have the same shape at runtime